draft-ietf-mls-protocol.md

```diff
@@ -3249,32 +3248,42 @@ struct {
 ~~~
 
 On receiving an MLSMessageContent containing a Proposal, a client MUST verify the
-signature inside MLSMessageAuth.  If the signature verifies
-successfully, then the Proposal should be cached in such a way that it can be
-retrieved by hash (as a ProposalOrRef object) in a later Commit message.
+signature inside MLSMessageAuth and that the `epoch` field of the enclosing
+MLSMessageContent is equal to the `epoch` field of the current GroupContext object.
+If the verification is successful, then the Proposal should be cached in such a way
+that it can be retrieved by hash (as a ProposalOrRef object) in a later Commit message.
 
 ### Add
 
 An Add proposal requests that a client with a specified KeyPackage be added
-to the group.  The proposer of the Add MUST verify the validity of the
-KeyPackage, as specified in {{keypackage-validation}}.
+to the group.
 
 ~~~ tls
 struct {
     KeyPackage key_package;
 } Add;
 ~~~
 
+An Add proposal is invalid if any of the following is true:
+
+* The KeyPackage is invalid according to {{keypackage-validation}}.
+
+* The LeafNode in the KeyPackage shares the same `signature_key` with another
+  LeafNode in the group.
+
+* The LeafNode in the KeyPackage shares the same `encryption_key` with another
+  LeafNode in the group.
+
+* The Credential in the KeyPackage represents a client already in the
+  group according to the application.
```

Comment:
I don't think this was agreed on

Reply:
This is how we interpreted the text
> If there are multiple Add proposals containing KeyPackages that the committer considers to represent the same client or a client already in the group (for example, identical KeyPackages or KeyPackages sharing the same Credential), the committer again chooses one to include and considers the rest invalid.

If this was not what was meant, the part "or a client already in the group" would be redundant. But of course I'll remove it if we were wrong.